File: jafu/src/test/java/org/springframework/fu/jafu/ApplicationDslTests.java
package org.springframework.fu.jafu;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.springframework.fu.jafu.Jafu.application;

import java.util.Locale;
import java.util.function.Consumer;

import org.junit.jupiter.api.Test;

import org.springframework.boot.WebApplicationType;
import org.springframework.boot.web.reactive.context.ReactiveWebServerApplicationContext;
import org.springframework.context.MessageSource;

class ApplicationDslTests {

	@Test
	void createAnEmptyApplicationAndCheckMessageSource() {
		var app = application(a -> {});
		var context = app.run();
		assertFalse(context instanceof ReactiveWebServerApplicationContext);
		var messageSource = context.getBean(MessageSource.class);
		assertEquals("Spring Fu!", messageSource.getMessage("sample.message", null, Locale.getDefault()));
		context.close();
	}

	@Test
	void createAnApplicationWithACustomBean() {
		var app = application(a -> a.beans(b -> b.bean(Foo.class)));
		var context = app.run();
		context.getBean(MessageSource.class);
		context.getBean(Foo.class);
		context.close();
	}

	@Test
	void createAnApplicationWithAConfigurationImport() {
		Consumer<ConfigurationDsl> conf = c -> c.beans(b -> b.bean(Foo.class));
		var app = application(a -> a.enable(conf));
		var context = app.run();
		context.getBean(MessageSource.class);
		context.getBean(Foo.class);
		context.close();
	}

	@Test
	void applicationProperties() {
		var app = application(a -> a.configurationProperties(City.class, "city"));
		var context = app.run();
		assertEquals(context.getBean(City.class).name, "San Francisco");
		context.close();
	}

	@Test
	void replaceBeanOfExistingApplication() {
		var app = application(a -> a.beans(b -> b.bean(Bar.class, () -> new Bar("original"))));
		var context = app.customize(a -> a.beans(b -> b.bean(Bar.class, () -> new Bar("customized"), c -> c.setPrimary(true)))).run();
		assertEquals("customized", context.getBean(Bar.class).value);
		context.close();
	}


	static class Foo {}

	static class Bar {

		private final String value;

		public Bar(String value) {
			this.value = value;
		}

		public String getValue() {
			return value;
		}
	}

	static class City {

		private String name;

		private String country;

		public String getName() {
			return name;
		}

		public void setName(String name) {
			this.name = name;
		}

		public String getCountry() {
			return country;
		}

		public void setCountry(String country) {
			this.country = country;
		}
	}

}

File: kofu/src/test/kotlin/org/springframework/fu/kofu/ApplicationDslTests.kt
/*
 * Copyright 2002-2018 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.springframework.fu.kofu

import io.mockk.every
import io.mockk.mockk
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.getBean
import org.springframework.boot.web.reactive.context.ReactiveWebServerApplicationContext
import org.springframework.context.MessageSource
import java.util.*

/**
 * @author Sebastien Deleuze
 */
class ApplicationDslTests {

	@Test
	fun `Create an empty application and check message source`() {
		val app = application { }
		with(app.run()) {
			assertFalse(this is ReactiveWebServerApplicationContext)
			val messageSource = getBean<MessageSource>()
			assertEquals("Spring Fu!", messageSource.getMessage("sample.message", null, Locale.getDefault()))
			close()
		}
	}

	@Test
	fun `Create an application with a custom bean`() {
		val app = application {
			beans {
				bean<Foo>(isPrimary = true)
			}
		}
		with(app.run()) {
			getBean<MessageSource>()
			getBean<Foo>()
			close()
		}
	}

	@Test
	fun `Create an application with a configuration import`() {
		val beanConfig = configuration {
			beans {
				bean<Foo>()
			}
		}
		val app = application {
			enable(beanConfig)
		}
		with(app.run()) {
			getBean<MessageSource>()
			getBean<Foo>()
			close()
		}
	}

	@Test
	fun `Mock a bean of an existing application`() {
		val app = application {
			beans {
				bean { Bar("original") }
			}
		}
		app.customize {
			beans {
				bean(isPrimary = true) {
					mockk<Bar> {
						every { value } returns "customized"
					}
				}
			}
		}
		with(app.run()) {
			assertEquals("customized", getBean<Bar>().value)
			close()
		}
	}


	class Foo
	class Bar(val value: String)

}
